[Clang][Diagnostics] Mention 'import std' in typeid diagnostic

AltriaSuki opened the pull request:
Previously, the diagnostic only suggested including `<typeinfo>`. Since C++20,the standard library may also be made available via `import std;`.

This change updates the diagnostic to mention `import std` as an alternative and adds a test to cover the new wording.

Commit:
[Clang][Diagnostics] Mention 'import std' in typeid diagnostic

AltriaSuki committed:
Merge branch 'main' into add-typeid-diagnostic

cor3ntin reviewed (commented):
Thanks for working on this.

We should be consistent (see the diagnostic about `<new>` in the next line, but there are many more.

More importantly, `import std` is only viable in some situations - in C++20 and greater with the appropriate config.

So while i agree it would be neat to suggest that option,  we should find out the cases in which it is a valid suggestion

ChuanqiXu9 reviewed (commented):
In the commit message we should say "Close https://github.com/llvm/llvm-project/issues/172966"

And also, yes, we can only do this when C++20 Modules are available. We can get this by LangOpts.CPlusPlus20. And also it will be good to do this to other diagnostics. e.g., <new> or anything else.

Commit:
[Clang][Diagnostics] Suggest 'import std' in typeid diagnostic for C++20+

  In C++20 and later, when typeid is used without including <typeinfo>,
  the diagnostic now suggests either including <typeinfo> or importing std,
  since modules are available in C++20.

  For C++17 and earlier versions, the diagnostic continues to only suggest
  including <typeinfo>, as modules are not available in those standards.

  Close #172966

AltriaSuki committed:
Merge branch 'main' into add-typeid-diagnostic

ChuanqiXu9 reviewed (commented):
LGTM generally. You can do the same thing for other headers, as `<new>` or `<coroutine>` in other PR.

AltriaSuki committed:
[Clang][Diagnostics] Mention 'import std' in typeid diagnostic

Update the typeid missing-<typeinfo> diagnostic to also mention import std
when compiling in C++20 and later. Use a %select clause to avoid duplicating
two nearly-identical diagnostics, and pass the selector from Sema.

Update typeid-requires-typeinfo.cpp to verify the C++20 vs
C++17 wording.

Close #172966

AltriaSuki committed:
Merge branch 'main' into add-typeid-diagnostic

cor3ntin reviewed (approved):
LGTM, thanks

cor3ntin commented:
Do you need us to merge that for you?

AltriaSuki commented:
> Do you need us to merge that for you?
Yes, that would be great. Thanks!
